hub/hub.go

```diff
@@ -4,80 +4,186 @@ import (
 	"fmt"
 	"log"
 	"net/http"
+	"net/url"
+	"time"
 
+	"github.com/dgrijalva/jwt-go"
 	"github.com/spf13/viper"
 	"go.uber.org/zap"
 )
 
-// Hub stores channels with clients currently subscribed and allows to dispatch updates.
-type Hub struct {
-	config             *viper.Viper
-	logger             Logger
-	transport          Transport
-	server             *http.Server
-	metricsServer      *http.Server
-	topicSelectorStore *TopicSelectorStore
-	metrics            *Metrics
+const HS256 = "HS256"
+
+// Option instances allow to configure the library.
+type Option func(h *Hub)
```

Comment:
This is a little bit problematic. Someone could easily modify the `Hub` struct after initialization:

```go
hub := New()

hub.Start()

WithDebug()(hub)
```

This can lead to unintended consequences. (This is long overdue to be fixed in my post)

One trick you can do is this:

```go
// Option instances allow to configure the library.
type Option option

type option func(h *Hub)
```

This way the API is not public, only the types are.

The other thing you can do is moving all configuration into an unexported struct:

```go
// Option instances allow to configure the library.
type Option func(o *options)

type options struct {
    debug bool
}
```

It allows you to somewhat separate the configuration itself from the main Hub. For example you could have a `newOptions` function in a separate file that creates a new options struct with defaults.

Reply:
fixed using a new struct, thanks for the tip.